walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,95 @@
+        allAttendees.addAll(request.getAttendees());
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        List<TimeRange> timeRangesIngoreOptional = new ArrayList<>();
+        List<TimeRange> timeRangesIncludeOptional = new ArrayList<>();
+        for (Event item : events) {
+            for (String s : request.getAttendees()) {
+                if (item.getAttendees().contains(s)) {
+                    timeRangesIngoreOptional.add(item.getWhen());
+                    break;
+                }
+            }
+            for (String s : allAttendees) {
+                if (item.getAttendees().contains(s)) {
+                    timeRangesIncludeOptional.add(item.getWhen());
+                    break;
+                }
+            }
+        }
+
+        Collection<TimeRange> ignoreOptional = availableTimeRange(timeRangesIngoreOptional, request);
+        Collection<TimeRange> includeOptional = availableTimeRange(timeRangesIncludeOptional, request);
+        if (request.getAttendees().size() == 0) {
+            return includeOptional;
+        }
+        if (includeOptional.size() != 0) {
+            return includeOptional;
+        }
+        return ignoreOptional;
+    }
+
+    // return timeRanges that satisfy meeting duration from all available timeRanges
+    public Collection<TimeRange> availableTimeRange(List<TimeRange> timeRanges, MeetingRequest request) {
+        Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+        int len = timeRanges.size();
+        if (len == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        if (len == 1) {
+            return Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timeRanges.get(0).start(), false),
+                TimeRange.fromStartEnd(timeRanges.get(0).end(), TimeRange.END_OF_DAY, true));
+        }
+        TimeRange pre = timeRanges.get(0);
+        int index = 1;
+        // Merge timerange that overlap
+        while (index < len) {
+            if (index == timeRanges.size()) {
+                break;
+            }
+            if (pre.overlaps(timeRanges.get(index))) {
+                int start = pre.start();
+                int end = Math.max(pre.end(), timeRanges.get(index).end());
+                timeRanges.remove(index);
+                timeRanges.remove(index - 1);
+                timeRanges.add(index - 1, TimeRange.fromStartEnd(start, end, false));
+                pre = timeRanges.get(index - 1);
+                len -= 1;
+            } else {
+                index += 1;
+            }
+        }
+            List<TimeRange> queries = new ArrayList<>();
+            int start = TimeRange.START_OF_DAY;
+        for (int i = 0; i < timeRanges.size(); i++) {
+            if (timeRanges.get(i).start() - start >= request.getDuration()) {
+                queries.add(TimeRange.fromStartEnd(start, timeRanges.get(i).start(), false));
+            }
+            start = timeRanges.get(i).end();
```

Comment:
I feel like you don't need to merge the overlapping time ranges if you change this line to `start = Math.max(start, timeRanges.get(i).end());`